File: Assignment 26 Feb Twitter live/src/main/java/com/pg/PalashGarg007/service/BookService.java
package com.pg.PalashGarg007.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.pg.PalashGarg007.pojo.Book;

import java.io.File;
import java.io.IOException;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

@Service
public class BookService {
	private static final String FILE_PATH = "books.json";
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    public List<Book> getAllBooks(String genre) {
        try {
            List<Book> books = objectMapper.readValue(
            	new File(FILE_PATH),
            	new TypeReference<List<Book>>() {}
            );

            if (genre != null && !genre.isEmpty()) {
                return books.stream()
                        .filter(book -> book.getGenre().equalsIgnoreCase(genre))
                        .collect(Collectors.toList());
            }
            
            return books;
            
        } catch (IOException e) {
            throw new RuntimeException("Error reading books from JSON file", e);
        }
        
    }
    
    public void saveBooks(List<Book> books) {
        try {
            objectMapper.writeValue(new File(FILE_PATH), books);
        } catch (IOException e) {
            throw new RuntimeException("Error writing books to JSON file", e);
        }
    }
    
    public Book getBookById(String id) {
        try {
            List<Book> books = objectMapper.readValue(
            	new File(FILE_PATH),
            	new TypeReference<List<Book>>() {}
            );

            if (id != null && !id.isEmpty()) {
                books = books.stream()
                        .filter(book -> book.getId().equalsIgnoreCase(id))
                        .collect(Collectors.toList());
            }
            
            return (books.isEmpty()) ? null : books.getFirst();
            
        } catch (IOException e) {
            throw new RuntimeException("Error reading books from JSON file", e);
        }
        
    }

    public void addBook(Book newBook) {
        List<Book> books = getAllBooks(null);
        books.add(newBook);
        saveBooks(books);
    }

    public boolean updateBookRating(String id, double newRating) {
        List<Book> books = getAllBooks(null);

        Optional<Book> bookOptional = books.stream()
                .filter(book -> book.getId().equals(id))
                .findFirst();

        if (bookOptional.isPresent()) {
            bookOptional.get().setRating(newRating);
            saveBooks(books);
            
            return true;
        }
        
        return false;
    }
    
    public Map<String, Object> getStatistics() {
        List<Book> books = getAllBooks(null);
        
        if (books.isEmpty()) {
            return Map.of("message", "No books available for statistics.");
        }

        Map<String, Double> avgRatingByGenre = books.stream()
                .collect(Collectors.groupingBy(Book::getGenre,
                        Collectors.averagingDouble(Book::getRating)));

        Book oldestBook = books.stream()
                .min(Comparator.comparingInt(Book::getPublicationYear))
                .orElse(null);

        Book newestBook = books.stream()
                .max(Comparator.comparingInt(Book::getPublicationYear))
                .orElse(null);

        Map<String, Object> statistics = new HashMap<>();
        statistics.put("averageRatingByGenre", avgRatingByGenre);
        statistics.put("oldestBook", oldestBook);
        statistics.put("newestBook", newestBook);
        
        return statistics;
    }

}

File: Assignment 26 Feb Twitter live/abhinow01/src/services/bookService.js
const fs = require('fs');
const path = require('path');

const dataPath = path.join(__dirname, '../data/books.json');

const readBooksFromFile = () => {
    try {
        const data = fs.readFileSync(path.join(__dirname, '../data/books.json'), 'utf-8');
        return JSON.parse(data); 
    } catch (error) {
        console.error("Error reading books file:", error);
        return []; 
    }
};

const writeBooksToFile = (books) => {
    fs.writeFileSync(dataPath, JSON.stringify(books, null, 2));
};

const getAllBooks = (genre) => {
    let books = readBooksFromFile();
    if (genre) {
        books = books.filter(book => book.genre.toLowerCase() === genre.toLowerCase());
    }
    return books;
};

const getBookById = (id) => {
    const books = readBooksFromFile();
    return books.find(book => book.id === id);
};

const addBook = (book) => {
    const books = readBooksFromFile();
    books.push(book);
    writeBooksToFile(books);
    return book;
};

const updateBookRating = (id, rating) => {
    const books = readBooksFromFile();
    const bookIndex = books.findIndex(book => book.id === id);

    if (bookIndex === -1) return null;

    books[bookIndex].rating = rating;
    writeBooksToFile(books);
    return books[bookIndex];
};

const getStatistics = async () => {
    let books =  readBooksFromFile();
    if (!books || books.length === 0) {
        return { message: "No books available for statistics" };
    }

    const avgRatingByGenre = books.reduce((acc, book) => {
        acc[book.genre] = acc[book.genre] || { sum: 0, count: 0 };
        acc[book.genre].sum += book.rating;
        acc[book.genre].count += 1;
        return acc;
    }, {});

    for (const genre in avgRatingByGenre) {
        avgRatingByGenre[genre] = (avgRatingByGenre[genre].sum / avgRatingByGenre[genre].count).toFixed(2);
    }

    const sortedByYear = books.sort((a, b) => a.publicationYear - b.publicationYear);

    return {
        averageRatingByGenre: avgRatingByGenre,
        oldestBook: sortedByYear[0],
        newestBook: sortedByYear[sortedByYear.length - 1]
    };
};

module.exports = {
    getAllBooks,
    getBookById,
    addBook,
    updateBookRating,
    getStatistics
};
